plugins/image-prioritizer/class-image-prioritizer-background-image-styled-tag-visitor.php

```diff
@@ -65,33 +79,126 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
+			 */
+			$lcp_element_external_background_image = $url_metric->get( 'lcpElementExternalBackgroundImage' );
+			if ( ! is_array( $lcp_element_external_background_image ) ) {
+				return null;
+			}
+			if ( null !== $previous_lcp_element_external_background_image && $previous_lcp_element_external_background_image !== $lcp_element_external_background_image ) {
+				return null;
+			}
+			$previous_lcp_element_external_background_image = $lcp_element_external_background_image;
+		}
+
+		return $previous_lcp_element_external_background_image;
+	}
+
+	/**
+	 * Maybe preloads external background image.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Context.
+	 */
+	private function maybe_preload_external_lcp_background_image( OD_Tag_Visitor_Context $context ): void {
+		// Gather the tuples of URL Metric group and the common LCP element external background image.
+		if ( ! is_array( $this->group_common_lcp_element_external_background_images ) ) {
+			$this->group_common_lcp_element_external_background_images = array();
+			foreach ( $context->url_metric_group_collection as $group ) {
+				$common = $this->get_common_lcp_element_external_background_image( $group );
+				if ( is_array( $common ) ) {
+					$this->group_common_lcp_element_external_background_images[] = array( $group, $common );
+				}
+			}
+		}
+
+		// There are no common LCP background images, so abort.
+		if ( count( $this->group_common_lcp_element_external_background_images ) === 0 ) {
+			return;
+		}
+
+		$processor = $context->processor;
+		$tag_name  = strtoupper( (string) $processor->get_tag() );
+		foreach ( $this->group_common_lcp_element_external_background_images as $i => list( $group, $common ) ) {
+			if (
+				// Note that the browser may send a lower-case tag name in the case of XHTML or embedded SVG/MathML, but
+				// the HTML Tag Processor is currently normalizing to all upper-case. The HTML Processor on the other
+				// hand may return the expected case.
+				strtoupper( $common['tag'] ) === $tag_name
+				&&
+				$processor->get_attribute( 'id' ) === $common['id'] // May be checking equality with null.
+				&&
+				$processor->get_attribute( 'class' ) === $common['class'] // May be checking equality with null.
+			) {
+				$this->add_preload_link( $context->link_collection, $group, $common['url'] );
+
+				// Now that the preload link has been added, eliminate the entry to stop looking for it while iterating over the rest of the document.
+				unset( $this->group_common_lcp_element_external_background_images[ $i ] );
+			}
+		}
+	}
+
+	/**
+	 * Adds an image preload link for the group.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Link_Collection  $link_collection Link collection.
+	 * @param OD_URL_Metric_Group $group           URL Metric group.
+	 * @param non-empty-string    $url             Image URL.
+	 */
+	private function add_preload_link( OD_Link_Collection $link_collection, OD_URL_Metric_Group $group, string $url ): void {
```

Comment:
Consider adding this as a public method on `OD_Link_Collection` directly (then without the first parameter of course), wrapping its `add_link()` method.

Regardless of whether it remains here or there, I think it would be good to rename it to `add_image_preload_link()`, as it's only for images but that's not obvious from the method name.

Reply:
Good idea. This is closely related to the direction that https://github.com/WordPress/performance/pull/1707 is going as well.

Reply:
I've just renamed it for now in fbb607664d7198fe70b8f535be42efa4820b0a59.

It's more complicated to make a general method for image preloading because a link may not have an `href` at all in the case of responsive preloads in which `imagesrcset` and `imagesizes` would be used instead (or in addition). It's simpler here for the background image case because it's just a single possible URL value as the `href`.

I'll keep this in mind for a future enhancement for facilitating image preloads, perhaps as part of https://github.com/WordPress/performance/pull/1707 assuming this PR is merged first, or else a separate PR.